Draft/RFC0001-PowerShell-Based-LCM.md

```diff
@@ -0,0 +1,92 @@
+---
+RFC: 0001
+Author: Ryan Bartram
+Status: Draft
+Version: 1.0
+Area: Engine
+Comments Due: 14/11/2019
+Plan to implement: <Yes | No>
+---
+
+# PowerShell Based LCM
+
+The Local Configuration Manager (LCM) was a windows component made available with the
+release of the WMF 4. It's the engine behind Desired State Configuration (DSC) and is used
+predominently as a Windows based configuration management tool.
+
+The current implementation is based on WMI/OMI and receives MOF files either from a Pull Server
+or by use of the cmdlets `Start-DscConfiguration` or `Publish-DscConfiguration`.
+
+The LCM can be configured to pull its configuration on a recurring basis and implements security
+by verifying signatures of MOFs or modules before they are downloaded.
+
+The LCM has 3 application modes `ApplyOnly`, `ApplyAndMonitor` and `ApplyAndAutoCorrect`. When
+the configuration is applied, it is applied in a serial fashion i.e. one resource after another.
+There is a dependency model defined within the configuration which enforces the order with which
+resources are deployed.
+
+## Motivation
+
+As a user, I want to replace the existing WMI/OMI based LCM on Windows and Linux with cross
+platform version that can perform all of the tasks of the existing LCM
+plus a lot lot more. So I can utilise my existing DSC knowledge to manage infrastructure
+all over my organisation. Onprem and in the cloud.
+
+I see this being the new Terraform, Polumi etc
+
+## Specification
+
+### User Experience
+
+```PowerShell
+# Command to generate new DSC MOF file
+Build-DscConfiguration
+
+# Command to execute DSC on a machine
+Invoke-DscConfiguration
+```
+
+### The basic Set
+This RFC proposes a way of using PowerShell to perform the task of the LCM:
+
+```PowerShell
+param([Resources[]]$resources)
+
+# Loop each resource and use the Invoke-DscResource cmdlet available in both Windows Powershell
+# and PowerShell Core to start each resource.
+
+foreach($resource in $resources) {
+    $InvokeParameters = @{
+        name = $resource.name
+        module = $resource.module
+        properties = $resource.properties
+    }
+
+    if ($false -eq (Invoke-DscResource @InvokeParameters -Method Test)) {
+        Invoke-DscResource @InvokeParameters -Method Set
```

Comment:
Yes, although there's the downside of some missing bits in the current implementation:
- missing PsDscRunAsCredential support (because windows & Linux have different impersonation models)
- DSC (and resources) still relying on CIM but CIM cmdlets not existing on Linux

There are ways to work around this (probably quickly mentioned in the RFC comments), and ideally this should filter through to `PSDesiredStateConfiguration.psm1` to keep it consistent with Implementations.

We should, for instance, create a command proxy for `Invoke-DscResource` that supports `Invoke-DscResourceAs` and if on windows and PsDscRunAsCredential is set, use a runspace and impersonate the provided user. If on Linux and PsDscRunAsCredential is set, start a PS Job as that user.

Reply:
I like the idea of an Invoke-DscResourceAs. I will add this to the RFC.

I haven't tested it yet (just breifly looked at the DSC Linux repo), but since DSC resources are just python scripts. We could import them and run the Get/test/set_Marshall commands.

Reply:
`Invoke-DscResource` in PS7 only supports PowerShell type of resource (MOF & Classes), and platforms (as long as the resource works on that platform).
I think we should not try to implement our own Python resource 'invocation', it will eventually be done by `Invoke-DscResource`, if there's value (and that should be open sourced at some point anyway). `Invoke-DscResourceAs` should just be a proxy command that adds the `PsDscRunAsCredential` support, by abstracting the underlying impersonation model (either with threadJobs & Impersonation in Windows or PS Jobs in Unix/Linux).